File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2025/Y25D04.kt
package org.gristle.pdxpuzzles.advent.y2025

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.objects.toMutableGrid

class Y25D04(private val input: String) : Day {
    private fun solve(loop: Boolean): Int {
        val grid = input.toMutableGrid { c -> c == '@' }
        var totalRemoved = 0
        do {
            val removable = grid.withIndex()
                .filter { (idx, b) ->
                    if (b) {
                        grid.getNeighbors(idx, true).count { it } < 4
                    } else {
                        false
                    }
                }
            for ((idx, _) in removable) {
                grid[idx] = false
            }
            totalRemoved += removable.size
        } while (loop && removable.isNotEmpty())
        return totalRemoved
    }

    override fun part1() = solve(false)
    override fun part2() = solve(true)
}

fun main() = Day.runDay(Y25D04::class)

//    Class creation: 1ms
//    Part 1: 1604 (22ms)
//    Part 2: 9397 (90ms)
//    Total time: 114ms

File: rust/advent/src/bin/2025/y25d04.rs
use advent::utilities::get_input::get_input;
use utilities::structs::{grid::Grid2, stopwatch::{ReportDuration, Stopwatch}};

type Input<'a> = &'a str;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(25, 4).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn solve(input: Input, run_until_stable: bool) -> usize {
    let mut grid = Grid2::new2d_map_str(input, |c| c == '@').unwrap();
    let mut total_removed = 0;
    loop {
        let removable = (0..grid.len())
            .filter(|&idx| {
                if grid[idx] {
                    let movable = grid.adjacent(idx, true)
                        .unwrap()
                        .filter(|adj| *adj.value)
                        .count() < 4;
                    if run_until_stable && movable {
                        grid[idx] = false;
                    }
                    movable
                } else {
                    false
                }
            })
            .count();
        total_removed += removable;
        if !run_until_stable || removable == 0 {
            break;
        }
    }
    total_removed
}

fn part1(notes: Input) -> Output {
    solve(notes, false)
}

fn part2(notes: Input) -> Output {
    solve(notes, true)
}

#[test]
fn default() {
    let input = get_input(25, 4).unwrap();
    assert_eq!(1604, part1(&input));
    assert_eq!(9397, part2(&input));
}

#[test]
fn test1() {
    let input = r"..@@.@@@@.
@@@.@.@.@@
@@@@@.@.@@
@.@@@@..@.
@@.@@@@.@@
.@@@@@@@.@
.@.@.@.@@@
@.@@@.@@@@
.@@@@@@@@.
@.@.@@@.@.
";
    assert_eq!(13, part1(&input));
}

#[test]
fn test2() {
    let input = r"..@@.@@@@.
@@@.@.@.@@
@@@@@.@.@@
@.@@@@..@.
@@.@@@@.@@
.@@@@@@@.@
.@.@.@.@@@
@.@@@.@@@@
.@@@@@@@@.
@.@.@@@.@.
";
    assert_eq!(43, part2(&input));
}

// Input parsed (24μs)
// 1. 1604 (1.134ms)
// 2. 9397 (12.946ms)
// Total: 14.108ms